pallets/basic-sto/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+type Settlement<T> = pallet_settlement::Module<T>;
+
+pub trait Trait:
+    frame_system::Trait + CommonTrait + IdentityTrait + pallet_settlement::Trait
+{
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+/// Details about the Fundraiser
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fundraiser<Balance> {
+    /// Token to raise funds in
+    pub raise_token: Ticker,
+    /// Amount of offering token available for sale
+    pub remaining_amount: Balance,
+    /// Price of one million offering token units (one full token) in terms of raise token units
+    pub price_per_token: u128,
+    /// Id of the venue to use for this fundraise
+    pub venue_id: u64,
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        Balance = <T as CommonTrait>::Balance,
+    {
+        /// A new fundraiser has been created
+        /// (offering token, raise token, amount to sell, price, venue id, fundraiser_id)
+        FundraiserCreated(IdentityId, Ticker, Ticker, Balance, u128, u64, u64),
+        /// An investor invested in the fundraiser
+        /// (offering token, raise token, offering_token_amount, raise_token_amount, fundraiser_id)
+        FundsRaised(IdentityId, Ticker, Ticker, Balance, Balance, u64),
+    }
+);
+
+decl_error! {
+    /// Errors for the Settlement module.
+    pub enum Error for Module<T: Trait> {
+        /// Sender does not have required permissions
+        Unauthorized,
+        /// An arithmetic operation overflowed
+        Overflow,
+        /// Not enough tokens left for sale
+        InsufficientTokensRemaining
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as StoCapped {
+        /// All fundraisers that are currently running. (ticker, fundraiser_id) -> Fundraiser
+        Fundraisers get(fn fundraisers): double_map hasher(blake2_128_concat) Ticker, hasher(twox_64_concat) u64 => Fundraiser<T::Balance>;
+        /// Total fundraisers created for a token
+        FundraiserCount get(fn fundraiser_count): map hasher(twox_64_concat) Ticker => u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Create a new offering.
+        #[weight = 200_000_000]
+        pub fn create_fundraiser(
+            origin,
+            offering_token: Ticker,
+            raise_token: Ticker,
+            sell_amount: T::Balance,
+            price_per_token: u128,
+            venue_id: u64
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::is_owner(&offering_token, did), Error::<T>::Unauthorized);
+            // TODO: Take custodial ownership of $sell_amount of $offering_token from treasury?
```

Comment:
I guess it will be the primaryIssuanceDid rather than the assetIssuer that runs the STO - maybe we should be checking that the origin's identity is the primaryIssuanceDid of the ticker, rather than the issuer?